src/orange/orangeinp/PolySolid.cc

```diff
@@ -0,0 +1,252 @@
+#endif
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+//! Construct the unioned "interior" of a polysolid
+template<class T>
+[[nodiscard]] NodeId construct_segments(PolySolidBase const& base,
+                                        T&& build_region,
+                                        detail::VolumeBuilder& vb)
+{
+    std::string const label{base.label()};
+    auto const& segments = base.segments();
+    CELER_ASSERT(segments.z().size() == segments.size() + 1);
+
+    SoftEqual soft_eq{vb.tol().rel};
+    std::vector<NodeId> segment_nodes;
+
+    for (auto i : range(segments.size()))
+    {
+        // Translate this segment along z
+        auto const [zlo, zhi] = segments.z(i);
+        if (soft_eq(zlo, zhi))
+        {
+            // Effectively zero height segment (degenerate: perhaps stacked
+            // cylinders, for example)
+            continue;
+        }
+        real_type const hz = (zhi - zlo) / 2;
+        auto scoped_transform
+            = vb.make_scoped_transform(Translation{{0, 0, zlo + hz}});
+
+        // Build outer shape
+        NodeId segment_node;
+        {
+            auto outer = build_region(segments.outer(i), hz);
+            segment_node = build_intersect_region(
+                vb, std::string{label}, std::to_string(i) + ".interior", outer);
+        }
+
+        if (segments.has_exclusion())
+        {
+            // Build inner shape
+            auto inner = build_region(segments.inner(i), hz);
+            NodeId inner_node = build_intersect_region(
+                vb, std::string{label}, std::to_string(i) + ".excluded", inner);
+
+            // Subtract (i.e., "and not") inner shape from this segment
+            auto sub_node = vb.insert_region({}, Negated{inner_node});
+            segment_node
+                = vb.insert_region(Label{label, std::to_string(i)},
+                                   Joined{op_and, {segment_node, sub_node}});
+        }
+        segment_nodes.push_back(segment_node);
+    }
+
+    // Union the given segments to create a new CSG node
+    return vb.insert_region(Label{label, "segments"},
+                            Joined{op_or, std::move(segment_nodes)});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct an enclosed angle if applicable.
+ */
+[[nodiscard]] NodeId construct_enclosed_angle(PolySolidBase const& base,
+                                              detail::VolumeBuilder& vb,
+                                              NodeId result)
+{
+    if (auto const& sea = base.enclosed_angle())
+    {
+        // The enclosed angle is "true" (specified by the user to truncate the
+        // shape azimuthally): construct a wedge to be added or deleted
+        auto&& [sense, wedge] = sea.make_wedge();
+        NodeId wedge_id
+            = build_intersect_region(vb, base.label(), "angle", wedge);
```

Comment:
Trying to understand the little differences:  why not ".wedge" instead of "angle"?
```suggestion
            = build_intersect_region(vb, base.label(), ".wedge", wedge);
```

Reply:
The `.` gets joined anyway; the extra `.` above is because it additionally has the integer segment inside it. So the surfaces will be `polycone.1.interior` or `polycone.angle`. The shape isn't always a wedge (it's sometimes an anti-wedge) and the name is consistent with the Solid.